node/src/neighborhood/gossip_acceptor.rs

```diff
@@ -134,40 +146,47 @@ debug! (self.logger, "IpChangeHandler qualifies() found a database Node with nod
-            let node_addr = db_node.node_addr_opt().expect ("A Node that accepts connections has no NodeAddr!");
-            for port in node_addr.ports() {
-                old_addrs.push (SocketAddr::new (node_addr.ip_addr(), port))
-            }
-            let new_node_addr = NodeAddr::new(&gossip_source.ip(), &node_addr.ports());
-            debug!(self.logger, "IpChange Gossip from {}; setting NodeAddr to: {:?}", gossip_source, new_node_addr);
-            db_node.metadata.node_addr_opt = Some (new_node_addr);
-            let local_addr = SocketAddr::from (database.root().node_addr_opt().as_ref()
-                .expect("Roots must have NodeAddr").clone());
-            old_addrs.into_iter().for_each (|old_addr| {
-                let msg = RemoveStreamMsg {
-                    local_addr,
-                    peer_addr: old_addr,
-                    stream_type: RemovedStreamType::Clandestine,
-                    sub: self.stream_shutdown_sub.clone(),
-                };
-                self.remove_stream_sub
-                    .try_send(msg)
-                    .expect ("StreamHandlerPool is dead");
-            });
-        }
-        else {
-            debug!(self.logger, "IpChange Gossip from {}; Originate-Only, no NodeAddr", gossip_source);
+            if let Some(node_addr) = db_node.metadata.node_addr_opt.as_ref() {
+                let old_addrs: Vec<SocketAddr> = node_addr.into();
+                let new_node_addr = NodeAddr::new(&gossip_source.ip(), &node_addr.ports());
+                debug!(
+                    self.logger,
+                    "IpChange Gossip from {}; setting NodeAddr to: {:?}",
+                    gossip_source,
+                    new_node_addr
+                );
+                db_node.metadata.node_addr_opt = Some(new_node_addr);
+                let local_addr = SocketAddr::from(
+                    database
+                        .root()
+                        .node_addr_opt()
+                        .as_ref()
+                        .expect("Roots must have NodeAddr"),
+                );
+                old_addrs.into_iter().for_each(|old_addr| {
+                    let msg = RemoveStreamMsg {
+                        local_addr,
+                        peer_addr: old_addr,
+                        stream_type: RemovedStreamType::Clandestine,
+                        sub: self.stream_shutdown_sub.clone(),
+                    };
+                    self.remove_stream_sub
+                        .try_send(msg)
+                        .expect("StreamHandlerPool is dead");
+                });
+            } else {
+                panic!("Node {:?} in the database sent Gossip, accepts connections, but has no NodeAddr!", db_node.public_key());
+            };
+            return GossipAcceptanceResult::Absorbed;
         }
-        GossipAcceptanceResult::Absorbed
+        GossipAcceptanceResult::Ignored
```

Comment:
A comment here or at the expression of the if condition `accept_connections`, saying this bears on originate only Nodes that _somehow_ (you know better than me why) can get in here. 

I wonder in the first place why originate only Nodes are able to send this IpChangeGossip at all? Wouldn't leaving them out serve with more utility than allowing it and then having to cope with it on this other side with certain difficulties?

Reply:
I fixed it.